doc/manual/calling-c-and-fortran-code.rst

```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+  + if ``T`` is an ``enum``, the argument type should be equivalent to ``Cint`` or ``Cuint``
+  + argument value will be copied (returned by-value)
+
+* ``struct T`` (including typedef to a struct)
+
+  + ``T``, where ``T`` is a Julia Leaf Type
+  + argument value will be copied (returned by-value)
+
+* ``void*``
+
+  + depends on how this parameter is used, first translate this to the intended pointer type,
+    then determine the Julia equivalent using the remaining rules in this list
+  + this argument may be declared as ``Ptr{Void}``, if it really is just an unknown pointer
+
+* ``jl_value_t*``
+
+  + ``Any``
+  + argument value must be a valid Julia object
+
+* ``jl_value_t**``
+
+  + ``Ref{Any}``
+  + argument value must be a valid Julia object (or ``C_NULL``)
+
+* ``T*``
+
+  + If the memory is already owned by Julia, or is an `isbits` type, and is known to be non-null:
+
+    + ``Ref{T}``, where ``T`` is the Julia type corresponding to ``T``
+    + a return type of ``Ref{Any}`` is invalid, it should either be ``Any``
+      (corresponding to ``jl_value_t*``) or ``Ptr{Any}`` (corresponding to ``Ptr{Any}``)
+    + currently partially unsupported by cfunction due to #2818
+    + C **MUST NOT** modify the memory returned via ``Ref{T}`` if ``T`` is an ``isbits`` type
+
+  + If the memory is owned by C:
+
+    + ``Ptr{T}``, where ``T`` is the Julia type corresponding to ``T``
+
+* ``(T*)(...)`` (e.g. a pointer to a function)
+
+  + ``Ptr{Void}`` (you may need to use ``cfunction`` explicitly to create this pointer)
 
 Passing Pointers for Modifying Inputs
--------------------------------------
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Because C doesn't support multiple return values, often C functions will take
-pointers to data that the function will modify. To accomplish this within a
-``ccall`` you need to encapsulate the value inside an array of the appropriate
-type. When you pass the array as an argument with a ``Ptr`` type, julia will
-automatically pass a C pointer to the encapsulated data::
+Because C doesn't support multiple return values,
+often C functions will take pointers to data that the function will modify.
+To accomplish this within a ``ccall``,
+you need to first encapsulate the value inside an ``Ref{T}`` of the appropriate type.
+When you pass this ``Ref`` object as an argument,
+julia will automatically pass a C pointer to the encapsulated data::
 
-    width = Cint[0]
-    range = Cfloat[0]
-    ccall(:foo, Void, (Ptr{Cint}, Ptr{Cfloat}), width, range)
+    width = Ref{Cint}(0)
+    range = Ref{Cfloat}(0)
+    ccall(:foo, Void, (Ref{Cint}, Ref{Cfloat}), width, range)
 
 This is used extensively in Julia's LAPACK interface, where an integer ``info``
 is passed to LAPACK by reference, and on return, includes the success code.
 
+Special Reference Syntax for ccall (deprecated):
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    The ``&`` syntax is deprecated, use the ``Ref{T}`` argument type instead
```

Comment:
+1 to no special syntax for this


> On Mar 10, 2015, at 7:22 PM, Jameson Nash <notifications@github.com> wrote:
> 
> In doc/manual/calling-c-and-fortran-code.rst:
> 
> >  
> >  This is used extensively in Julia's LAPACK interface, where an integer ``info``
> >  is passed to LAPACK by reference, and on return, includes the success code.
> >  
> > +Special Reference Syntax for ccall (deprecated):
> > +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> > +
> > +    The ``&`` syntax is deprecated, use the ``Ref{T}`` argument type instead
> scratch that. i just tried it, and it would potentially actually require removing the special parsing of & to make that possible. then you could define:
> 
> &(A) = Ref(A)
> &(A::Type) = Ref{A}
> but i don't really recommend abusing and replacing the bitwise-and operator in this way.
> 
> i went with a type-based approach precisely because it avoided needing special-syntax to define ccall (which would have transferred over to ccall). It should be easier now to work with the AST (since there aren't Expr(:&) nodes to deal with) and easier to automatically generate ccall/cfunction expressions (which also was the motivation for attempting to minimize the number of places that codegen branches on isbits/bitstype/leaftype, rather than syntactic expressions).
> 
> —
> Reply to this email directly or view it on GitHub.
>